sql/api/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -117,65 +130,144 @@ object JavaTypeInference {
+      // should not consider class as bean, as otherwise it will be treated as empty schema
+      // and loose the data on deser.
+      if (properties.isEmpty && seenTypeSet.nonEmpty) {
+        if (classOf[KryoSerializable].isAssignableFrom(c)) {
+          Encoders.kryo(c).asInstanceOf[AgnosticEncoder[_]]
+        } else if (classOf[java.io.Serializable].isAssignableFrom(c)) {
+          Encoders.javaSerialization(c).asInstanceOf[AgnosticEncoder[_]]
+        } else {
+          throw ExecutionErrors.cannotFindEncoderForTypeError(t.getTypeName)
+        }
+      } else {
+        // add type variables from inheritance hierarchy of the class
+        val parentClassesTypeMap =
+          JavaTypeUtils.getTypeArguments(c, classOf[Object]).asScala.toMap
+        val classTV = parentClassesTypeMap ++ typeVariables
+        // Note that the fields are ordered by name.
+        val fields = properties.map { property =>
+          val readMethod = property.getReadMethod
+          val methodReturnType = readMethod.getGenericReturnType
+          val encoder = encoderFor(methodReturnType, seenTypeSet + c, classTV)
+          // The existence of `javax.annotation.Nonnull`, means this field is not nullable.
+          val hasNonNull = readMethod.isAnnotationPresent(classOf[Nonnull])
+          EncoderField(
+            property.getName,
+            encoder,
+            encoder.nullable && !hasNonNull,
+            Metadata.empty,
+            Option(readMethod.getName),
+            Option(property.getWriteMethod).map(_.getName))
+        }
+        // implies it cannot be assumed a BeanClass.
+        // Check if its super class or interface could be represented by an Encoder
+
+        JavaBeanEncoder(ClassTag(c), fields.toImmutableArraySeq)
       }
-      JavaBeanEncoder(ClassTag(c), fields.toImmutableArraySeq)
 
-    case _ =>
-      throw ExecutionErrors.cannotFindEncoderForTypeError(t.toString)
+    case t =>
+      throw ExecutionErrors.cannotFindEncoderForTypeError(t.getTypeName)
   }
 
+  private def getAllSuperClasses(typee: Type): Array[Class[_]] = Option(typee)
+    .map({
+      case clazz: Class[_] =>
+        val currentInterfaces = clazz.getInterfaces
+        currentInterfaces ++ (clazz.getSuperclass +: currentInterfaces).flatMap(clz =>
+          getAllSuperClasses(clz)) :+ clazz
+
+      case tv: TypeVariable[_] =>
+        tv.getBounds.foldLeft(Array.empty[Class[_]]) { case (res, bnd) =>
+          res ++ getAllSuperClasses(bnd)
+        }
+    })
+    .getOrElse(Array.empty)
+
   def getJavaBeanReadableProperties(beanClass: Class[_]): Array[PropertyDescriptor] = {
     val beanInfo = Introspector.getBeanInfo(beanClass)
     beanInfo.getPropertyDescriptors
       .filterNot(_.getName == "class")
       .filterNot(_.getName == "declaringClass")
       .filter(_.getReadMethod != null)
   }
+
+  object UseSerializationEncoder {
+    def unapply(th: Throwable): Option[Class[_] => AgnosticEncoder[_]] = th match {
```

Comment:
Can we use the actual superclasses here instead of going through the error message? I'd also prefer if you unify this with the other java serialization code.

Reply:
Unifying it will other java serialziation code, is something which I intend to do in my refactoring..

Reply:
I had thought about using the Serializable encoders as part of the match/case statement, but it appears tricky as the idea is that the Serializable encoders are to be used as last resort.  And the way current logic works ( if I am not wrong), it collects all the interfaces/classes via the CommonUtils method, and it is possible , I think, that Serializable interface may show early as the introspection data  is converted to a map in the Bean class

        val parentClassesTypeMap =
          JavaTypeUtils.getTypeArguments(c, classOf[Object]).asScala.toMap

I am hesitant at this point to further complicate the code, unless you all think that its worthwhile to do it in this PR.